src/main/java/reposense/report/ReportGenerator.java

```diff
@@ -46,34 +50,67 @@ public static void generateReposReport(List<RepoConfiguration> configs, String o
         InputStream is = RepoSense.class.getResourceAsStream(TEMPLATE_FILE);
         FileUtil.copyTemplate(is, outputPath);
 
-        cloneAndAnalyzeRepos(configs, outputPath);
+        Map<RepoLocation, List<RepoConfiguration>> repoLocationMap = groupConfigsByRepoLocation(configs);
+        cloneAndAnalyzeRepos(repoLocationMap, outputPath);
 
         FileUtil.writeJsonFile(new SummaryReportJson(configs, generationDate), getSummaryResultPath(outputPath));
         logger.info("The report is generated at " + outputPath);
     }
 
     /**
-     * Clone, analyze and generate the report for repositories in {@code configs}.
+     * Groups {@code RepoConfiguration} with the same {@code RepoLocation} together so that they are only cloned once.
+     */
+    private static Map<RepoLocation, List<RepoConfiguration>> groupConfigsByRepoLocation(
+            List<RepoConfiguration> configs) {
+        Map<RepoLocation, List<RepoConfiguration>> repoLocationToBranch = new HashMap<>();
+        for (RepoConfiguration config : configs) {
+            RepoLocation location = config.getLocation();
+            if (repoLocationToBranch.containsKey(location)) {
+                repoLocationToBranch.get(location).add(config);
+            } else {
+                List<RepoConfiguration> configsList = new ArrayList<>();
+                configsList.add(config);
+                repoLocationToBranch.put(location, configsList);
+            }
+        }
+        return repoLocationToBranch;
+    }
+
+    /**
+     * Clone, analyze and generate the report for repositories in {@code repoLocationMap}.
      * Performs analysis and report generation of each repository in parallel with the cloning of the next repository.
      */
-    private static void cloneAndAnalyzeRepos(List<RepoConfiguration> configs, String outputPath) throws IOException {
+    private static void cloneAndAnalyzeRepos(
+            Map<RepoLocation, List<RepoConfiguration>> repoLocationMap, String outputPath) throws IOException {
         RepoCloner repoCloner = new RepoCloner();
-        RepoConfiguration clonedRepo = null;
+        RepoLocation clonedRepoLocation = null;
 
-        for (RepoConfiguration config : configs) {
-            repoCloner.clone(outputPath, config);
+        for (RepoLocation location : repoLocationMap.keySet()) {
+            repoCloner.clone(outputPath, repoLocationMap.get(location).get(0));
```

Comment:
We already have a RepoLocation. Maybe we can use the location to clone the repo directly.

Reply:
Ah, the reason I use RepoConfiguration instead of just RepoLocation is because during cloning I need some properties of RepoConfiguration, for example repoRoot folder, displayName, etc.